### PR TITLE
Optimize column listing to not require fetching table properties

### PR DESCRIPTION
**Pull request by hashhar**
## Description

Before this change table metadata was fetched when listing columns which
ended up fetching table properties for JDBC connectors which incurs I/O
in connectors which support them.

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# ClickHouse, SQL Server
* Improve performance of listing columns. ({issue}`23429`)
```

### CODE REVIEW

**Comment by hashhar** [comment]
while working on this I noticed that we do confusing things in some places.

e.g. `ConnectorMetadata#getTableMetadata#getColumns` excludes columns like `$path` in Hive connector but `ConnectorMetadata#getColumnHandles` includes them. I guess this is to prevent "materializing" those hidden columns in other SQL operations but this makes for confusion. And this is also what prevents applying this optimisation to non-JDBC connectors (or even JDBC connectors too in future if they start returning different columns from `getTableMetadata#getColumns` VS `getColumnHandles`.

**Comment by hashhar** [comment]
For Ignite and Phoenix specifically I'll send a follow-up PR to "unhide" the autogenerated columns we add in case the user doesn't specify primary key (which is required for those systems).

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlJdbcConnectionCreation.java [hunk 1]
```diff
@@ -62,7 +62,9 @@ protected QueryRunner createQueryRunner()
                     runner.createCatalog("counting_postgresql", "counting_postgresql", ImmutableMap.of(
                             "connection-url", postgreSqlServer.getJdbcUrl(),
                             "connection-user", postgreSqlServer.getUser(),
-                            "connection-password", postgreSqlServer.getPassword()));
+                            "connection-password", postgreSqlServer.getPassword(),
+                            // connection reuse hides the actual number of I/O operations done
```
Comment (kokosing) [inline plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlJdbcConnectionCreation.java:66]: When this test was written, it was the case to see that reuse reuses the connection. Even test name stresses that "ConnectionCreation", here what you do is metadata call counting. Which is a bit different, maybe you need a different test?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java [hunk 2]
```diff
@@ -1082,7 +1082,13 @@ public Map<SchemaTableName, List<ColumnMetadata>> listTableColumns(ConnectorSess
         for (SchemaTableName tableName : tables) {
             try {
                 jdbcClient.getTableHandle(session, tableName)
-                        .ifPresent(tableHandle -> columns.put(tableName, getTableMetadata(session, tableHandle).getColumns()));
+                        .ifPresent(tableHandle -> {
```
Comment (kokosing) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:1085]: please maybe extract some method. Nested lambas that are using outer variables are hard to follow.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java [hunk 3]
```diff
@@ -1082,7 +1082,13 @@ public Map<SchemaTableName, List<ColumnMetadata>> listTableColumns(ConnectorSess
...
+                            List<ColumnMetadata> columnMetadata = getColumnHandles(session, tableHandle).values()
+                                    .stream()
+                                    .map(columnHandle -> getColumnMetadata(session, tableHandle, columnHandle))
```
Comment (kokosing) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:1088]: You can cast to `JdbcColumnHandle`

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlJdbcConnectionCreation.java [hunk 4]
```diff
@@ -26,22 +26,18 @@
 import io.trino.plugin.jdbc.JdbcPlugin;
```
[the remark this answers is not in the document]
Comment (hashhar) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlJdbcConnectionCreation.java:26]: you're right, there's very little reuse. the reused parts are:
- testing session setup
- tpch plugin + catalog install
- handling of closing of query runner

It's mostly about having to not create query runner from scratch and instead use existing builder. I'll update commit message to what you proposed, makes more sense.

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlJdbcConnectionCreation.java (hunk 1 above)
Comment (hashhar) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlJdbcConnectionCreation.java:66]: Good point. I did check and there was `TestReusableConnectionFactory` but that's only a unit test.

**Comment by hashhar** [comment]
@kokosing PTAL at the fixups + [Add tests to count I/O operations in JDBC connectors](https://github.com/trinodb/trino/pull/23429/commits/58d674fbc4cbe2c6a72d6f31a9a810f8c2df81e9) (which is the only new commit, it replaces [Disable connection reuse in connection counting tests](https://github.com/trinodb/trino/pull/23429/commits/29ca2e63c6b07d7d67ab11275b9ef58ff9d58fd1)).

**Review by kokosing — APPROVED** [review]
Well done!
